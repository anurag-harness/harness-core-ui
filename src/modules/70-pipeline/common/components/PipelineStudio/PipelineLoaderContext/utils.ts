import { DefaultNewPipelineId, RouteState } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineActions'
import { EntityGitDetails } from 'services/cd-ng'
import { getId, getRepoIdentifierName } from '@pipeline/components/PipelineStudio/PipelineContext/utils'
import { IDB } from '@modules/10-common/components/IDBContext/IDBContext'

export const deletePipelineFromIDB = async <T = unknown>({
  routeState,
  identifier,
  gitDetails,
  idb
}: {
  routeState: RouteState
  identifier: string
  gitDetails?: EntityGitDetails
  idb: IDB<T>
}): Promise<void> => {
  const id = getId(
    routeState.accountIdentifier,
    routeState.orgIdentifier || '',
    routeState.projectIdentifier || '',
    identifier,
    getRepoIdentifierName(gitDetails),
    gitDetails?.branch || ''
  )
  await idb.del(id)

  const defaultId = getId(
    routeState.accountIdentifier,
    routeState.orgIdentifier || '',
    routeState.projectIdentifier || '',
    DefaultNewPipelineId,
    getRepoIdentifierName(gitDetails),
    gitDetails?.branch || ''
  )
  await idb.del(defaultId)
}
