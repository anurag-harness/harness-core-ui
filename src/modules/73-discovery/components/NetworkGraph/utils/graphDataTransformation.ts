/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { Node, Edge } from 'reactflow'
import { NodeTypes, nodeGroupOptions, nodeOptions } from '@discovery/components/NetworkGraph/constants'
import type {
  ApiCreateNetworkMapRequest,
  ApiListDiscoveredServiceConnection,
  ApiListDiscoveredService
} from 'services/servicediscovery'
import type {
  NetworkMapEdgeData,
  NetworkMapNodeData,
  ServiceEdgeData,
  ServiceNodeData
} from '@discovery/components/NetworkGraph/types'

export function getGraphNodesFromServiceList(
  serviceList: ApiListDiscoveredService | null
): Node<ServiceNodeData, NodeTypes>[] {
  if (!serviceList) return []

  const namespaces = new Set<string>()
  const graphNodes: Node<ServiceNodeData, NodeTypes>[] = []
  serviceList.items?.map(service => {
    const namespace = service.spec.kubernetes?.namespace
    if (namespace && !namespaces.has(namespace)) {
      namespaces.add(namespace)
    }

    if (service.id)
      graphNodes.push({
        ...nodeOptions,
        id: service.id,
        data: { ...service, name: service.spec.kubernetes?.name ?? 'service' },
        parentNode: namespace
      })
  })

  namespaces.forEach(value => {
    graphNodes.push({
      ...nodeGroupOptions,
      id: value,
      data: { id: value, name: value }
    })
  })

  return graphNodes
}

export function getGraphEdgesFromServiceConnections(
  connectionList: ApiListDiscoveredServiceConnection | null
): Edge<ServiceEdgeData>[] {
  if (!connectionList) return []

  const graphEdges: Edge<ServiceEdgeData>[] = []

  connectionList.items?.map(connection => {
    if (connection.id && connection.sourceID && connection.destinationID)
      graphEdges.push({
        id: connection.id,
        source: connection.sourceID,
        target: connection.destinationID,
        data: { parentNode: connection.sourceNamespace, ...connection }
      })
  })

  return graphEdges
}

export function getGraphNodesFromNetworkMap(
  networkMap: ApiCreateNetworkMapRequest | undefined
): Node<NetworkMapNodeData, NodeTypes>[] {
  if (!networkMap) return []

  const namespaces = new Set<string>()
  const graphNodes: Node<NetworkMapNodeData, NodeTypes>[] = []

  networkMap.resources?.map(service => {
    const namespace = service.kubernetes?.namespace
    if (namespace && !namespaces.has(namespace)) {
      namespaces.add(namespace)
    }

    if (service.id)
      graphNodes.push({
        ...nodeOptions,
        type: NodeTypes.NetworkMapHexagon,
        id: service.id,
        data: service,
        parentNode: namespace
      })
  })

  namespaces.forEach(value => {
    graphNodes.push({
      ...nodeGroupOptions,
      id: value,
      data: { id: value, name: value, kind: 'discoveredservice' }
    })
  })

  return graphNodes
}

export function getGraphEdgesFromNetworkMap(
  networkMap: ApiCreateNetworkMapRequest | undefined
): Edge<NetworkMapEdgeData>[] {
  if (!networkMap) return []

  const graphEdges: Edge<NetworkMapEdgeData>[] = []

  networkMap.connections?.map(connection => {
    if (connection.from?.id && connection.to?.id)
      graphEdges.push({
        id: `${connection.from.id}-${connection.to.id}`,
        source: connection.from.id,
        target: connection.to.id,
        data: { ...connection, parentNode: connection.from.kubernetes?.namespace }
      })
  })

  return graphEdges
}
