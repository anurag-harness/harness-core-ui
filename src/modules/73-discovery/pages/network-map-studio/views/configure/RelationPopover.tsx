/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { IPopoverProps, Position } from '@blueprintjs/core'
import {
  Button,
  ButtonVariation,
  FormInput,
  Formik,
  FormikForm,
  Layout,
  MultiTypeInputType,
  Popover,
  SelectOption,
  Text
} from '@harness/uicore'
import { Color, FontVariation, Intent } from '@harness/design-system'
import * as Yup from 'yup'
import React from 'react'
import { useStrings } from 'framework/strings'
import {
  ApiCreateNetworkMapRequest,
  DatabaseConnection,
  DatabaseConnectionType,
  DatabaseNetworkMapEntity
} from 'services/servicediscovery'
import MultiTypeMapInputSet from '@modules/70-pipeline/components/InputSetView/MultiTypeMapInputSet/MultiTypeMapInputSet'

export interface RelationPopoverProps {
  children: React.ReactElement
  networkMap: ApiCreateNetworkMapRequest
  updateNetworkMap: (networkMap: ApiCreateNetworkMapRequest) => Promise<void>
  isOpen: boolean
  open: () => void
  close: () => void
  popoverProps?: IPopoverProps
  initialValues?: ServiceRelationFormType
}

export interface ServiceRelationFormType {
  source: string
  target: string
  properties: Record<string, string>
}

export default function RelationPopover({
  children,
  networkMap,
  updateNetworkMap,
  isOpen,
  open,
  close,
  popoverProps,
  initialValues = { source: '', target: '', properties: { type: 'TCP', port: '' } }
}: RelationPopoverProps): React.ReactElement {
  const { getString } = useStrings()

  const servicesSelectOptions = React.useMemo(
    () => networkMap.resources?.map(s => ({ label: s.name, value: s.id } as SelectOption)) ?? [],
    [networkMap]
  )

  return (
    <Popover
      isOpen={isOpen}
      onInteraction={nextOpenState => {
        if (nextOpenState) open()
        else close()
      }}
      position={Position.RIGHT_TOP}
      modifiers={{ preventOverflow: { escapeWithReference: true } }}
      canEscapeKeyClose={false}
      minimal
      {...popoverProps}
    >
      {children}
      <Formik<ServiceRelationFormType>
        formName="newServiceRelationForm"
        initialValues={initialValues}
        validationSchema={Yup.object().shape({
          source: Yup.string().required(
            getString('common.validation.fieldIsRequired', {
              name: getString('discovery.sourceService')
            })
          ),
          target: Yup.string().required(
            getString('common.validation.fieldIsRequired', {
              name: getString('discovery.targetService')
            })
          ),
          properties: Yup.object()
            .shape({
              type: Yup.string()
                .equals(['TCP', 'UDP'])
                .required(
                  getString('common.validation.fieldIsRequired', {
                    name: getString('typeLabel')
                  })
                ),
              port: Yup.number().required(
                getString('common.validation.fieldIsRequired', {
                  name: getString('common.smtp.port')
                })
              )
            })
            .required(
              getString('common.validation.fieldIsRequired', {
                name: getString('discovery.relationProperties')
              })
            )
        })}
        onSubmit={async values => {
          let source: DatabaseNetworkMapEntity | undefined
          let destination: DatabaseNetworkMapEntity | undefined
          networkMap.resources.map(service => {
            if (service.id === values.source) source = service
            if (service.id === values.target) destination = service
          })

          const modifiedConnections = networkMap.connections?.filter(
            conn => values.source !== conn.from?.id || values.target !== conn.to?.id
          )
          const { type, port, ...params } = values.properties

          const newConnection: DatabaseConnection = {
            from: source,
            to: destination,
            type: type as DatabaseConnectionType,
            manual: true, // because this is only used for custom connections
            port,
            params
          }

          modifiedConnections?.push(newConnection)

          const newNetworkMap: ApiCreateNetworkMapRequest = {
            ...networkMap,
            connections: modifiedConnections
          }
          await updateNetworkMap(newNetworkMap)
          close()
        }}
      >
        {formik => {
          return (
            <FormikForm>
              <Layout.Vertical spacing="medium" padding="medium">
                <Text font={{ variation: FontVariation.CARD_TITLE }}>
                  {getString('discovery.serviceRelationshipDetails')}
                </Text>
                <Layout.Horizontal spacing="medium">
                  <FormInput.Select
                    name="source"
                    items={servicesSelectOptions}
                    label={getString('discovery.sourceService')}
                  />
                  <FormInput.Select
                    name="target"
                    items={servicesSelectOptions}
                    label={getString('discovery.targetService')}
                  />
                </Layout.Horizontal>
                <div>
                  <MultiTypeMapInputSet
                    appearance={'minimal'}
                    cardStyle={{ width: '50%' }}
                    name="properties"
                    valueMultiTextInputProps={{
                      allowableTypes: [MultiTypeInputType.FIXED]
                    }}
                    multiTypeFieldSelectorProps={{
                      label: (
                        <Text font={{ variation: FontVariation.SMALL_BOLD }}>
                          {getString('discovery.relationProperties')}
                        </Text>
                      ),
                      disableTypeSelection: true,
                      allowedTypes: [MultiTypeInputType.FIXED],
                      hideError: false
                    }}
                    formik={formik}
                  />
                  {Object.entries(formik.errors.properties ?? {}).map(([key, error]) => (
                    <Text
                      key={key}
                      intent={Intent.DANGER}
                      icon="circle-cross"
                      iconProps={{ size: 12, color: Color.RED_600 }}
                    >
                      {error}
                    </Text>
                  ))}
                </div>
                <Layout.Horizontal spacing="medium">
                  <Button type="submit" variation={ButtonVariation.PRIMARY} text={getString('saveChanges')} />
                  <Button
                    variation={ButtonVariation.TERTIARY}
                    text={getString('cancel')}
                    onClick={async () => {
                      await updateNetworkMap({ ...networkMap })
                      close()
                    }}
                  />
                </Layout.Horizontal>
              </Layout.Vertical>
            </FormikForm>
          )
        }}
      </Formik>
    </Popover>
  )
}
