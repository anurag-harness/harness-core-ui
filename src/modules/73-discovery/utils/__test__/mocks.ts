/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { DatabaseNetworkMapEntity } from 'services/servicediscovery'

export const connectionsBetweenServices = [
  {
    from: {
      id: '64920dc166c663ba792cf3b0',
      kind: 'discoveredservice',
      name: 'access-control',
      kubernetes: { namespace: 'chaos-1000' }
    },
    port: '9090',
    to: {
      id: '64920dc166c663ba792cf3bb',
      kind: 'discoveredservice',
      name: 'harness-manager',

      kubernetes: { namespace: 'chaos-1000' }
    },
    type: 'TCP',
    manual: false
  },
  {
    from: {
      id: '64920dc166c663ba792cf3bb',
      kind: 'discoveredservice',
      name: 'access-control',

      kubernetes: { namespace: 'chaos-1000' }
    },
    port: '6379',
    to: {
      id: '64920dc166c663ba792cf3b0',
      kind: 'discoveredservice',
      name: 'redis-sentinel-harness-announce-0',

      kubernetes: { namespace: 'chaos-1000' }
    },
    type: 'TCP',
    manual: false
  }
]

export const mockConnections = {
  items: [
    {
      id: '64920e3866c663ba792cf569',
      type: 'TCP',
      sourceID: '64920dc166c663ba792cf3b0',
      sourceName: 'access-control',
      sourceNamespace: 'chaos-1000',
      sourceIP: '10.100.2.2',
      destinationID: '64920dc166c663ba792cf3bb',
      destinationName: 'harness-manager',
      destinationNamespace: 'chaos-1000',
      destinationIP: '10.104.11.119',
      destinationPort: '9090'
    },
    {
      id: '64920e3866c663ba792cf56a',
      type: 'TCP',
      sourceID: '64920dc166c663ba792cf3bb',
      sourceName: 'access-control',
      sourceNamespace: 'chaos-1000',
      sourceIP: '10.100.2.2',
      destinationID: '64920dc166c663ba792cf3b0',
      destinationName: 'redis-sentinel-harness-announce-0',
      destinationNamespace: 'chaos-1000',
      destinationIP: '10.104.0.162',
      destinationPort: '6379'
    }
  ]
}

export const mockServices = {
  items: [
    {
      id: '64920dc166c663ba792cf3b0',
      agentID: '64920dbb66c663ba792cf134',
      kind: 'Service',
      name: 'access-control',
      namespace: 'chaos-1000',
      uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c89',
      service: {
        owner: {
          kind: 'Service',
          namespace: 'chaos-1000',
          name: 'access-control',
          uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c89',
          apiVersion: 'v1'
        },
        ports: [
          {
            name: 'http',
            protocol: 'TCP',
            port: 9006,
            targetPort: 9006
          }
        ],
        clusterIP: '10.104.11.160',
        clusterIPs: ['10.104.11.160'],
        externalIPs: undefined,
        loadBalancerIP: '',
        externalName: '',
        type: 'ClusterIP'
      },
      createdAt: '2023-06-20T20:36:17.94Z',
      updatedAt: '2023-06-20T20:36:18.357Z',
      createdBy: '',
      updatedBy: '',
      removed: false
    },
    {
      id: '64920dc166c663ba792cf3bb',
      agentID: '64920dbb66c663ba792cf134',
      kind: 'Service',
      name: 'chaos-exporter',
      namespace: 'chaos-1000',
      uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c88',
      service: {
        owner: {
          kind: 'Service',
          namespace: 'chaos-1000',
          name: 'chaos-exporter',
          uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c89',
          apiVersion: 'v1'
        },
        ports: [
          {
            name: 'http',
            protocol: 'TCP',
            port: 9006,
            targetPort: 9006
          }
        ],
        clusterIP: '10.104.11.160',
        clusterIPs: ['10.104.11.160'],
        externalIPs: undefined,
        loadBalancerIP: '',
        externalName: '',
        type: 'ClusterIP'
      },
      createdAt: '2023-06-20T20:36:17.94Z',
      updatedAt: '2023-06-20T20:36:18.357Z',
      createdBy: '',
      updatedBy: '',
      removed: false
    }
  ],
  page: {
    all: true,
    index: 0,
    totalItems: 1,
    totalPages: 1
  }
}

export const mockNetworkMapResources: DatabaseNetworkMapEntity[] = [
  {
    id: '64920dc166c663ba792cf3b0',
    kind: 'discoveredservice',
    name: 'access-control',
    kubernetes: {
      namespace: 'chaos-1000'
    }
  },
  {
    id: '64920dc166c663ba792cf3bb',
    kind: 'discoveredservice',
    name: 'chaos-exporter',
    kubernetes: {
      namespace: 'chaos-1000'
    }
  }
]

export const mockLogData = `{"level":"","pos":1,"out":"{file:entry.go:41,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go version go1.20.8,time:2023-10-04T04:30:15.872344691Z}","time":"2023-10-04T04:30:20.480817808Z","args":{"PodName":"sd-cluster-9ts9x","PodNamespace":"hce"}}
{"level":"","pos":2,"out":"{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}","time":"2023-10-04T04:30:20.843837081Z","args":{"PodName":"sd-cluster-9ts9x","PodNamespace":"hce"}}
{"level":"","pos":2,"out":"{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}","time":"2023-10-04T04:30:20.843837081Z","args":{"PodName":"sd-node-2gc9l","PodNamespace":"hce"}}`

export const mockLogDataResponse = [
  {
    text: {
      level: '',
      out: '{file:entry.go:41,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go version go1.20.8,time:2023-10-04T04:30:15.872344691Z}',
      args: {
        PodName: 'sd-cluster-9ts9x',
        PodNamespace: 'hce'
      }
    }
  },
  {
    text: {
      level: '',
      out: '{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}',
      args: {
        PodName: 'sd-cluster-9ts9x',
        PodNamespace: 'hce'
      }
    }
  },
  {
    text: {
      args: {
        PodName: 'sd-node-2gc9l',
        PodNamespace: 'hce'
      },
      level: '',
      out: '{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}'
    }
  }
]

export const installations = {
  id: '651cea4d2dc5d1dc8a74530b',
  agentID: '65126c42183f27a2381f9384',
  delegateTaskID: '2c4sK0KkRsivFp6gb4kDRA-DEL',
  delegateID: '',
  delegateTaskStatus: 'SUCCESS',
  agentDetails: {
    cluster: {
      name: 'sd-cluster-9ts9x',
      namespace: 'hce',
      uid: 'c63fbf54-0f32-4f3a-ad54-078cf8bddf41',
      status: 'Succeeded'
    },
    node: [
      {
        name: 'sd-node-2gc9l',
        namespace: 'hce',
        uid: '234d86b8-454d-434f-87ff-5493e0f0f1ea',
        status: 'Succeeded'
      }
    ],
    status: 'SUCCESS'
  },
  isCronTriggered: true,
  logStreamID: 'c70fbd79-4f7d-45e7-b65c-df4b6237336f',
  logStreamCreatedAt: '2023-10-04T04:30:05.339Z',
  isLogStreamOpen: false,
  createdAt: '2023-10-04T04:30:05.339Z',
  updatedAt: '2023-10-04T04:45:15.299Z',
  createdBy: '',
  updatedBy: '',
  removed: false
}
