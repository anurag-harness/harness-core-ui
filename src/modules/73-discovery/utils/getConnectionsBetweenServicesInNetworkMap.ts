/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import {
  ApiListDiscoveredServiceConnection,
  DatabaseConnection,
  DatabaseConnectionType,
  DatabaseNetworkMapEntity
} from 'services/servicediscovery'

export default function getConnectionsBetweenServicesInNetworkMap(
  services: DatabaseNetworkMapEntity[] | null | undefined,
  connectionList: ApiListDiscoveredServiceConnection | null
): DatabaseConnection[] | undefined {
  // Checks for new connections between the selected services
  const serviceHashMap = new Map<string, DatabaseNetworkMapEntity>(services?.map(s => [s.id ?? '', s]))

  const connectionBetweenSelectedServices: DatabaseConnection[] | undefined = connectionList?.items
    ?.filter(conn => serviceHashMap.has(conn.sourceID ?? '') && serviceHashMap.has(conn.destinationID ?? ''))
    .map(conn => ({
      from: {
        id: conn.sourceID,
        kind: serviceHashMap.get(conn.sourceID ?? '')?.kind ?? 'discoveredservice',
        name: conn.sourceName,
        kubernetes: {
          namespace: conn.sourceNamespace
        }
      },
      port: conn.destinationPort,
      to: {
        id: conn.destinationID,
        kind: serviceHashMap.get(conn.destinationID ?? '')?.kind ?? 'discoveredservice',
        name: conn.destinationName,
        kubernetes: {
          namespace: conn.destinationNamespace
        }
      },
      type: conn.type as DatabaseConnectionType,
      manual: false
    }))

  return connectionBetweenSelectedServices
}
