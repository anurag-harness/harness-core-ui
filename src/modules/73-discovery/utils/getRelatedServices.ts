/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import {
  ApiListDiscoveredServiceConnection,
  ApiListDiscoveredService,
  DatabaseDiscoveredServiceCollection
} from 'services/servicediscovery'

export function getRelatedServices(
  serviceID: string,
  serviceList: ApiListDiscoveredService | null,
  connectionList: ApiListDiscoveredServiceConnection | null
): DatabaseDiscoveredServiceCollection[] | undefined {
  if (!serviceList || !connectionList) return

  const relatedServiceIds = connectionList.items?.map(conn => {
    if (serviceID === conn.sourceID) return conn.destinationID ?? ''
    /* istanbul ignore next */
    if (serviceID === conn.destinationID) return conn.sourceID ?? ''
  })

  return serviceList?.items?.filter(service => service.id === serviceID || relatedServiceIds?.includes(service.id))
}
