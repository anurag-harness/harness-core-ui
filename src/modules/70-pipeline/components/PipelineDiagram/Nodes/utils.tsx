/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isNumber } from 'lodash-es'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import { dragPlaceholderImageBase64 } from './assets/dragImageBase64'
import { dragStagePlaceholderImageBase64 } from './assets/dragStageImageBase64'
import defaultCss from './DefaultNode/DefaultNode.module.scss'

interface DotNotationPathProps {
  identifier: string
  entityType?: string
  baseDotNotation?: string
}

export enum NodeEntity {
  STAGE = 'STAGE',
  STEP = 'STEP'
}
export enum NodeWrapperEntity {
  step = 'step',
  stepGroup = 'stepGroup',
  stage = 'stage'
}

export const COLLAPSED_MATRIX_NODE_LENGTH = 8
export const MAX_ALLOWED_MATRIX_COLLAPSED_NODES = 4
export const DEFAULT_MATRIX_PARALLELISM = 1

export const getPositionOfAddIcon = (props: any, isRightNode?: boolean): string => {
  if (isRightNode) {
    return '-40px'
  }
  if (props?.children?.length) {
    if (props?.prevNode?.children?.length) {
      return '-65px'
    }
    if (props?.prevNode) return '-58px'
  }
  if (props?.prevNode?.children?.length) {
    return '-58px'
  }
  if (props?.parentIdentifier && !props.prevNode) {
    return '-35px'
  }
  return '-50px'
}

export const transformMatrixLabels = (nodeData: string): string => {
  const parsedNodeName = {}
  try {
    // if object parse to yaml
    const parsedVal = JSON.parse(nodeData)
    if (parsedVal && !isNumber(parsedVal)) {
      Object.assign(parsedNodeName, parsedVal)
      return yamlStringify(parsedNodeName, { indent: 4 })
    } else {
      return JSON.stringify(nodeData)
    }
  } catch (_e) {
    // name is string/number, parse it to string
    return JSON.stringify(nodeData)
  }
}

export const attachDragImageToEventHandler = (event: React.DragEvent<HTMLDivElement>, type?: NodeEntity): void => {
  // set drag image preview to custom icon in case of safari browser, as safari blocks image preview if dom tree have transform property on any parent
  if (navigator.userAgent.search('Safari') >= 0 && navigator.userAgent.search('Chrome') < 0) {
    const dragImage = document.createElement('img')
    dragImage.src = type === NodeEntity.STAGE ? dragStagePlaceholderImageBase64 : dragPlaceholderImageBase64
    dragImage.style.position = 'absolute'
    dragImage.style.left = '-100%'

    document.body.append(dragImage)
    setTimeout(() => dragImage.remove(), 0)

    event.dataTransfer.setDragImage(dragImage, dragImage.width / 2, dragImage.height / 2)
  }
}

export const getEntityIdentifierBasedDotNotationPath = ({
  baseDotNotation,
  identifier,
  entityType = NodeWrapperEntity.step
}: DotNotationPathProps): string => {
  return `${baseDotNotation}.${entityType}.${identifier}`
}

export const getBaseDotNotationWithoutEntityIdentifier = (dotNotation = ''): string => {
  const lastDotIndex = dotNotation.lastIndexOf('.')
  return lastDotIndex !== -1 ? dotNotation.substring(0, lastDotIndex) : dotNotation
}

export const getConditionalClassName = (isDisabled: boolean, className?: string): { className?: string } => {
  return isDisabled ? { className: defaultCss.disabledIcon } : className ? { className } : {}
}
