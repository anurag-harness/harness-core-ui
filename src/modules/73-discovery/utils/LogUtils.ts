/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { sanitizeHTML } from '@common/utils/StringUtils'
import { formatDatetoLocale } from '@common/utils/dateUtils'
import { loggerFor } from 'framework/logging/logging'
import { ModuleName } from 'framework/types/ModuleName'
import { DatabaseInstallationCollection } from 'services/servicediscovery'

export interface ArgsType {
  PodName: string
  PodNamespace: string
}

export interface DAgentLogs {
  clusterLogs: string
  nodeLogs: string
}

/**
 * Data for a log unit/section
 */
export interface LogSectionData {
  /**
   * The log data
   */
  data: LogLineData[]
}

/**
 * Keys in the json from server
 */
export type TextKeys = 'level' | 'out' | 'time' | 'args'

export interface LogLineData {
  /**
   * Text for the line
   */
  text: Partial<Record<TextKeys, string | ArgsType>>
  /**
   * Indices of the search result with in the current line.
   * This will be a continous list
   *
   * If this values is `[3, 4]`, it means the result number 3 and 4 will be found in this line
   */
  searchIndices?: Partial<Record<TextKeys, number[]>>
}

const logger = loggerFor(ModuleName.CHAOS)

export function processLogsData(data: string): LogLineData[] {
  return String(data)
    .split('\n')
    .reduce<LogSectionData['data']>((accumulator, line) => {
      /* istanbul ignore else */
      if (line.length > 0) {
        try {
          const { level, time, out, args } = JSON.parse(line) as Record<string, string>
          accumulator.push({
            text: {
              level: sanitizeHTML(level),
              time: formatDatetoLocale(time),
              out: sanitizeHTML(out),
              args: args
            }
          })
        } catch (error) {
          logger.error(`Error while parsing logs: ${error}`)
        }
      }
      return accumulator
    }, [])
}

export function processAgentLogs(
  logData: string | undefined,
  installationList: DatabaseInstallationCollection | undefined
): DAgentLogs {
  const logsData = processLogsData(logData ?? '')
  let clusterLogs = ''
  let nodeLogs = ''
  const nodeNames: Array<string> = []
  installationList?.agentDetails?.node?.forEach(node => {
    if (node?.name) nodeNames.push(node?.name)
  })
  logsData.map(log => {
    if (installationList?.agentDetails?.cluster?.name === (log.text.args as ArgsType)?.PodName) {
      clusterLogs = clusterLogs + log.text.out + '\n'
    }
    if (nodeNames.includes((log.text.args as ArgsType)?.PodName)) {
      nodeLogs = nodeLogs + log.text.out + '\n'
    }
  })

  return {
    clusterLogs: clusterLogs.length ? clusterLogs : 'No cluster level logs found',
    nodeLogs: nodeLogs.length ? nodeLogs : 'No node level logs found'
  }
}
