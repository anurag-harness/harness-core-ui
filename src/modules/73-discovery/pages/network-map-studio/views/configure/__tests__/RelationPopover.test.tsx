/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import { accountPathProps, modulePathProps, networkMapPathProps, projectPathProps } from '@common/utils/routeUtils'
import { mockNetworkMap } from './mocks'
import RelationPopover from '../RelationPopover'

const PATH = routes.toCreateNetworkMap({
  ...accountPathProps,
  ...projectPathProps,
  ...modulePathProps,
  ...networkMapPathProps
})
const PATH_PARAMS = {
  accountId: 'accountId',
  orgIdentifier: 'default',
  projectIdentifier: 'Discovery_Test',
  module: 'chaos',
  dAgentId: 'dAgent-1',
  networkMapId: 'testnetworkmap'
}

const updateNetworkMap = jest.fn()
const openNewRelation = jest.fn()
const closeNewRelation = jest.fn()

describe('<RelationPopover /> tests with data', () => {
  beforeEach(() => jest.clearAllMocks())

  test('should render component and click on submit', async () => {
    const { container, getByText } = render(
      <TestWrapper path={PATH} pathParams={PATH_PARAMS}>
        <RelationPopover
          networkMap={mockNetworkMap}
          updateNetworkMap={updateNetworkMap}
          isOpen={true}
          open={openNewRelation}
          close={closeNewRelation}
          initialValues={{
            source: '64c229096fb4bc8feefc933b',
            target: '64c229096fb4bc8feefc933d',
            properties: { type: 'TCP', port: '3000', tag: 'hello' }
          }}
        >
          <div className="button"></div>
        </RelationPopover>
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()

    act(() => {
      fireEvent.click(getByText('saveChanges'))
    })
  })

  test('should render component show error and click on cancel', async () => {
    const { container, getByText } = render(
      <TestWrapper path={PATH} pathParams={PATH_PARAMS}>
        <RelationPopover
          networkMap={mockNetworkMap}
          updateNetworkMap={updateNetworkMap}
          isOpen={true}
          open={openNewRelation}
          close={closeNewRelation}
          initialValues={{
            source: '64c229096fb4bc8feefc933b',
            target: '64c229096fb4bc8feefc933d',
            properties: { type: 'notTCP', port: '30wd00', tag: 'hello' }
          }}
        >
          <div className="button"></div>
        </RelationPopover>
      </TestWrapper>
    )

    act(() => {
      fireEvent.click(getByText('saveChanges'))
    })

    act(() => {
      fireEvent.click(getByText('cancel'))
    })

    expect(container).toMatchSnapshot()
  })
})
