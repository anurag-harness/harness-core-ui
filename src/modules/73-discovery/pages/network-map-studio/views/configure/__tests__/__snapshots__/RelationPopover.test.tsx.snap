// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RelationPopover /> tests with data should render component and click on submit 1`] = `
<div>
  <span
    class="bp3-popover-wrapper"
  >
    <span
      class="bp3-popover-target bp3-popover-open"
    >
      <div
        class="button"
      />
    </span>
  </span>
</div>
`;

exports[`<RelationPopover /> tests with data should render component show error and click on cancel 1`] = `
<div>
  <span
    class="bp3-popover-wrapper"
  >
    <span
      class="bp3-popover-target bp3-popover-open"
    >
      <div
        class="button"
      />
    </span>
  </span>
</div>
`;
